api/ask_astro/slack/controllers/mention.py

```diff
@@ -51,13 +51,7 @@ async def send_answer(request: AskAstroRequest, say: AsyncSay, ts: str):
     """
     response = markdown_to_slack(request.response)
 
-    source_links = ", ".join(
-        f"<{link}|[{n_}]>"
-        for n_, link in enumerate(
-            {*[s.name for s in request.sources if s.name]},
-            start=1,
-        )
-    )
+    source_links = ", ".join(f"<{s.name}|[{n_}]>" for n_, s in enumerate(request.sources, start=1) if s.name)
```

Comment:
The previous logic ensures that we do not send duplicate docs due to the set operation. Do we not need to ensure that we do not send duplicate document sources?

Reply:
Correct me if I am wrong but this shouldn't be a problem. Because
1. When querying a vector DB, a single query it should never be able to pull duplicate entities. In the case where we literally ingested them as duplicates, then we have much bigger problems to worry about which is fixing our ingestion logic.
2. However, your concern is still valid since we do run multiple queries through prompt rewording using MultiQuery Retriever from LangChain, but they already took this concern into account here: https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/retrievers/multi_query.py#L215. Tada! When the pulled documents are put together, it removes duplicates! 

So it should be fine right now. I do agree that it may be a good idea to double check for duplicates downstream again though. In that case we may need to loop over the new list, but we still cannot use `set` since set will disrupt the original ordering of the documents.

Reply:
thanks @davidgxue for the explanation. will leave it to @sunank200 if he wishes to do another loop to remove duplicates (if any). 

Additionally, somewhere in the pipeline, should we check for duplicate ingested documents and raise an alert if found? I guess we could create a separate issue for that as you said it could be a bigger problem @davidgxue?

Reply:
Thanks for the comment @pankajkoti. This is the same implementation on the UI as well. We wanted to keep both UI and slack responses the same. 

During ingestion, we take care of duplicates in the weaviate provider.